src/uopool/services/sanity_check.rs

```diff
@@ -0,0 +1,559 @@
+
+        Ok(())
+    }
+
+    fn verification_gas(
+        &self,
+        user_operation: &UserOperation,
+    ) -> Result<(), BadUserOperationError<M>> {
+        if user_operation.verification_gas_limit > self.max_verification_gas {
+            return Err(BadUserOperationError::HighVerificationGasLimit {
+                verification_gas_limit: user_operation.verification_gas_limit,
+                max_verification_gas: self.max_verification_gas,
+            });
+        }
+
+        let calculated_pre_verification_gas =
+            Overhead::default().calculate_pre_verification_gas(user_operation);
+        if user_operation.pre_verification_gas < calculated_pre_verification_gas {
+            return Err(BadUserOperationError::LowPreVerificationGas {
+                pre_verification_gas: user_operation.pre_verification_gas,
+                calculated_pre_verification_gas,
+            });
+        }
+
+        Ok(())
+    }
+
+    async fn verify_paymaster(
+        &self,
+        user_operation: &UserOperation,
+        entry_point: &Address,
+    ) -> Result<(), BadUserOperationError<M>> {
+        if !user_operation.paymaster_and_data.is_empty() {
+            let paymaster_address = if user_operation.paymaster_and_data.len() >= 20 {
+                Address::from_slice(&user_operation.paymaster_and_data[0..20])
+            } else {
+                return Err(BadUserOperationError::PaymasterVerification {
+                    paymaster_and_data: user_operation.paymaster_and_data.clone(),
+                });
+            };
+
+            let mempool_id = mempool_id(entry_point, &self.chain_id);
+
+            if let Some(entry_point) = self.entry_points.get(&mempool_id) {
+                let deposit_info = entry_point
+                    .get_deposit_info(&paymaster_address)
+                    .await
+                    .map_err(|_| BadUserOperationError::PaymasterVerification {
+                        paymaster_and_data: user_operation.paymaster_and_data.clone(),
+                    })?;
+
+                if U256::from(deposit_info.deposit) > user_operation.max_fee_per_gas {
+                    if let Some(reputation) = self.reputations.read().get(&mempool_id) {
+                        if reputation.get_status(&paymaster_address) != ReputationStatus::BANNED {
+                            self.sanity_check_results
+                                .write()
+                                .entry(user_operation.hash(&entry_point.address(), &self.chain_id))
+                                .or_insert_with(Default::default)
+                                .insert(SanityCheckResult::PaymasterVerified);
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn call_gas_limit(
+        &self,
+        user_operation: &UserOperation,
+    ) -> Result<(), BadUserOperationError<M>> {
+        let non_zero_value_call = gas::non_zero_value_call();
```

Comment:
I thought it would require `eth_estimate` API.

Reply:
You are right, I will check this again. Maybe non zero value call is not even necessary.